presto-spi/src/main/java/io/prestosql/spi/block/ByteArrayBlockEncoding.java

```diff
@@ -50,15 +65,56 @@ public Block readBlock(BlockEncodingSerde blockEncodingSerde, SliceInput sliceIn
     {
         int positionCount = sliceInput.readInt();
 
-        boolean[] valueIsNull = decodeNullBits(sliceInput, positionCount).orElse(null);
-
+        byte[] valueIsNullPacked = retrieveNullBits(sliceInput, positionCount);
         byte[] values = new byte[positionCount];
-        for (int position = 0; position < positionCount; position++) {
-            if (valueIsNull == null || !valueIsNull[position]) {
-                values[position] = sliceInput.readByte();
+
+        if (valueIsNullPacked == null) {
+            sliceInput.readBytes(Slices.wrappedBuffer(values));
+            return new ByteArrayBlock(0, positionCount, null, values);
+        }
+        boolean[] valueIsNull = decodeNullBits(valueIsNullPacked, positionCount);
+
+        int nonNullPositionCount = sliceInput.readInt();
+        sliceInput.readBytes(Slices.wrappedBuffer(values, 0, nonNullPositionCount));
+        int position = nonNullPositionCount - 1;
+
+        // Handle Last (positionCount % 8) values
+        for (int i = positionCount - 1; i >= (positionCount & ~0b111) && position >= 0; i--) {
+            values[i] = values[position];
+            if (!valueIsNull[i]) {
+                position--;
             }
         }
 
+        // Handle the remaining positions.
+        for (int i = (positionCount & ~0b111) - 8; i >= 0 && position >= 0; i -= 8) {
```

Comment:
Maybe it's possible to extract this logic for single array code, e.g:
```
<T> void unpackValues(boolean[] valueIsNull, T values)
```
so that we don't have to copy this tricky code for every block

Reply:
It is not. Either we work with T[] which is an array of objects, not primitives, or with T and provide setters and getters lambdas which are based on objects as well. Both are performance suicides.
I can move this code to a single class with 5 almost identical methods and write Javadoc about it. WDYT?

Reply:
> I can move this code to a single class with 5 almost identical methods and write Javadoc about it. WDYT?

I don't think it matters much. You could do that as as follow-up